static/tutorial/exception-safety.md

```diff
@@ -0,0 +1,341 @@
+# Exception Safety
+
+This tutorial covers a relatively broad and previously underdocumented
+topic in the Haskell ecosystem. In particular, goals of this tutorial
+are:
+
+* Explain the different types of exceptions
+* Explain how exceptions can be generated
+* Explain how interruptible actions work
+    * More to the point: interruptible actions introduce a concept of
+      trading off between *correctness* and *deadlock prevention*
+* Define requirements for proper exception safety in
+    * Normal code
+    * Resource acquisition code
+    * Resource cleanup code
+* Give guidelines on how to write exception safe code in general
+* Point out some existing violations of these rules in the Haskell
+  ecosystem
+
+Many of the ideas behind this tutorial originated with the
+[safe-exceptions](/package/safe-exceptions) package, and as such it is
+strongly recommended that you work with that package to get maximum
+correctness and ease-of-use.
+
+Like the safe-exceptions package, this tutorial takes runtime
+exceptions in Haskell as a fact, not a point for discussion, as this
+is the reality of the library ecosystem today. No guidelines are given
+here for when you yourself should generate runtime exceptions versus
+using something like `IO (Either MyError a)` or `ExceptT`. That
+discussion is left for a different time.
+
+## Types of exceptions
+
+We're going to follow the safe-exceptions package's definitions for
+exception types. It is strongly recommended that you read
+[the safe-exceptions tutorials](/package/safe-exceptions) first. As a
+quick summary:
+
+* While there is a difference in the type hierarchy for synchronous
+  and asynchronous exceptions (via the `SomeAsyncException` type),
+  this distinction is not enforced by the `Control.Exception` module
+  (e.g., you can synchronously throw via `throwIO` an
+  asynchronously-typed exception).
+* Despite this limitation, we assume that we can fully distinguish
```

Comment:
I believe the safe-exception docs already cover this point. In sum: the vast majority of existing code follows these rules already. Technically speaking though, existing code that violates these assumptions could leak to unrecoverable exceptions being treated as recoverable or vice-versa. That would be bad, but far from catastrophic.

I don't know if it's worth adding any of this here, since we're just giving a short summary of a different document for framing.